registry/service.go

```diff
@@ -36,23 +36,23 @@ type Service interface {
 	LoadInsecureRegistries([]string) error
 }
 
-// DefaultService is a registry service. It tracks configuration data such as a list
+// defaultService is a registry service. It tracks configuration data such as a list
 // of mirrors.
-type DefaultService struct {
+type defaultService struct {
 	config *serviceConfig
 	mu     sync.Mutex
 }
 
-// NewService returns a new instance of DefaultService ready to be
+// NewService returns a new instance of defaultService ready to be
 // installed into an engine.
-func NewService(options ServiceOptions) (*DefaultService, error) {
+func NewService(options ServiceOptions) (Service, error) {
```

Comment:
I would argue that, while `DefaultService` values are not meant to be constructed outside the package, they are meant to be used. Proof: it satisfies the `Service` interface and has an exported constructor. [The Go community has coalesced on the recommendation:](https://github.com/golang/go/wiki/CodeReviewComments#interfaces)
> The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring.

Reply:
It's a good point, and looking at this interface again, I think I want to cleanup some of this. Generally if there's only a single implementation, I prefer returning (and using) concrete types, which also makes it easier to understand the code (instead of looking up "what implementation are we dealing with here?"). For the `registry.Service` interface;

- it's implemented by `registry.defaultService`
- it's implemented by `fakeService` in daemon/images; this is only for testing, and that doesn't need the full registry service, only `SearchRegistryForImages()`. I already was looking at splitting the "search" functionality from the "registry" service (as "search" is a deprecated V1 feature, and unrelated to anything else "registry"); some preparations for that are in https://github.com/moby/moby/pull/43312, but I have a branch on my machine to continue that work after.
- it's implemented by `pluginRegistryService` in `plugins`. From an initial look, it looks like that _only_ requires the `LookupPullEndpoints()`; we should consider having the `plugins` package define an interface for that (so that that package defines the contract).

Reply:
opened https://github.com/moby/moby/pull/43358 for the pluginRegistryService